### PR TITLE
feat: Implement declarative task dependencies in `fluid-build`

### PR DESCRIPTION
**Pull request by curtisman**
Instead of implicit trying determine dependencies by assuming certain way our package's build command are structured, we will now specify these `TaskDefinitions` with dependencies declaratively in the package's `package.json` fluidBuild properties in combination of the global fluidBuild config that will be inherited if not overwritten by the package's `package.json`. 
  - Packages following previous build structures (e.g. 'build:gen-ver', 'tsc', 'build:esnext', 'eslint', 'prettier') will have reasonable defaults in the global fluidBuild config and doesn' t need to specify anything in the package's `package.json`.
  - Packages has more flexible fto structure their build slightly differently, and still able to declare the order for the build correctly
  - The depenencies can be reasoned thru the config instead of reading the inplicit (and repo specific rules) in the fluid-build code.
  - The change fixes some existing build order issues in building PropertyDDS packages.
  - It will now correctly instantiate dependent tasks whether it is part of the script or not (e.g., `-t webpack`, will include the necessary `build:esnext` build now).

Repo policy check is implemented to enforce build dependencies that are specified in `tsconfig.json` and `.eslintrc.js`. Also clean up some `fluid-build`'s unnecessary check on  build related scripts now that there are no implicit rules that relies on build script structures.

NOTE:
- All the build task definition in the main repo are additive, and doesn't interfere with the old `fluid-build`.  So it is safe to check in now and be ready when the new `fluid-build` is released
- After this change is released and adopted, follow up to clean up old config and build rules (e.g. `build`, `build:compile`, `lint`, etc.) and replace them that calls to `fluid-build`

BREAKING CHANGE: The `--script` flag has been removed. Use the `--task` flag instead.

### CODE REVIEW

File: build-tools/packages/build-tools/README.md [hunk 1]
```diff
@@ -79,7 +79,7 @@ Options:
   -r --rebuild        Clean and build on matched packages (all if package regexp is not specified)
      --reinstall      Same as --uninstall --install.
      --root <path>    Root directory of the Fluid repo (default: env _FLUID_ROOT_)
-  -s --script <name>  npm script to execute (default:build)
+  -t --target <name>  target to execute (default:build)
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/README.md:82]: Do we have to change this?

File: build-tools/packages/build-tools/src/common/fluidTaskDefinitions.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+ */
+
+import { getFluidBuildConfig } from "./fluidUtils";
+import { PackageJson } from "./npmPackage";
+
+/**
+ * Task definitions (type `TypeDefinitions`) is an object describing build tasks for fluid-build.
+ * Task names are represented as property name on the object and the value the task configuration
+ * (type `TypeConfig`). Task configuration can a plain array of string, presenting the task's
+ * dependencies or a full description (type `TypeConfigFull`).
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/common/fluidTaskDefinitions.ts:13]:
```suggestion
 * Task definitions (type `TaskDefinitions`) is an object describing build tasks for fluid-build.
 * Task names are represented as property name on the object and the value the task configuration
 * (type `TaskConfig`). Task configuration can a plain array of string, presenting the task's
 * dependencies or a full description (type `TaskConfigFull`).
```

Did you mean this?

**Comment by tylerbutler** [comment]
So is the script parsing just gone altogether? Everything has to be explicitly configured now, either in the root or in the package.json? If so I am sad to see the parsing go away. It is the really innovative thing in fluid build imo.

**Comment by curtisman** [comment]
The capability remain (&&, npm run, concurrently).  See Task.Create. But the implicit dependencies are now done explicitly (See *Task.addDependencies)

In the old way,  
- every time there is a new rule or dependency,  we need to change the code to accommodate,  now we can do it by changing package.json.  
- the old way couldn't express more complicate dependencies, or isolate the dependencies chain (e.g.  it doesn't work if you want to only build build:wane t and able detect the task that is needed.)
- It is easier to parse and reason able the explicit dependencies.

File: build-tools/packages/build-tools/README.md (hunk 1 above)
Comment (curtisman) [reply build-tools/packages/build-tools/README.md:82]: I can make both work the same given that it doesn't necessarily need to be "script"

**Review by tylerbutler — APPROVED** [review]

**Pull request merged**